derived_types/main.cpp

```diff
@@ -0,0 +1,92 @@
+#include <iostream>
+
+void printArray(int arr[], const int size) {
+    for(int i = 0; i < size; i++) {
+        std::cout << arr[i] << " ";
+    };
+    std::cout << std::endl;
+}
+
+void selectionSort(int* arr, const int size) {
+    for(int i = 0; i < size; i++) {
+        int min = arr[i];
+        for(int j = i + 1; j < size; j++) {
+            if (min > arr[j])
+                min = arr[j];
+        }
+        std::swap(arr[i], min);
+    }
+}
+
+struct Complex {
+    double real;
+    double imaginary;
+};
+
+struct Point {
+    int x;
+    int y;
+};
+
+struct Rectangle {
+    Point p1;
+    Point p2;
+    Point p3;
+    Point p4;
+};
+
+void magicFunction(void* unknownStruct, const short flag) {
+    switch(flag) {
+        case 1: {
+            auto complex = (Complex*)unknownStruct;
```

Comment:
Let's use C++ constructions for type casting against legacy from C. It's more useable, because C++ construction will raise compilation error, but C-style cast can be verified only in runtime